# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [3]:
import pandas as pd

import numpy as np

df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv")

In [4]:
pd.set_option('display.max_columns', None) 

df.sample(2)

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
3356        3356  PS47550  California             13178.431630       No   
7786        7786  VK57280         NaN              4727.354548      NaN   

     Coverage             Education Effective To Date EmploymentStatus Gender  \
3356    Basic  High School or Below           1/13/11         Employed      F   
7786    Basic              Bachelor            1/3/11       Unemployed      M   

      Income Location Code Marital Status  Monthly Premium Auto  \
3356   80237      Suburban        Married                   108   
7786       0      Suburban         Single                    72   

      Months Since Last Claim  Months Since Policy Inception  \
3356                      4.0                             71   
7786                     23.0                              7   

      Number of Open Complaints  Number of Policies    Policy Type  \
3356                        0.0                   2  Personal Auto   
7786                        0.0                   3  Personal Auto   

           Policy Renew Offer Type Sales Channel  Total Claim Amount  \
3356  Personal L3           Offer1        Branch          776.601764   
7786  Personal L2           Offer2   Call Center          345.600000   

      Vehicle Class Vehicle Size Vehicle Type  
3356            SUV      Medsize            A  
7786  Four-Door Car      Medsize          NaN

In [5]:
df.columns = [col.lower().replace(" ", "_") for col in df.columns]

In [6]:
df.columns

Index(['unnamed:_0', 'customer', 'state', 'customer_lifetime_value',
       'response', 'coverage', 'education', 'effective_to_date',
       'employmentstatus', 'gender', 'income', 'location_code',
       'marital_status', 'monthly_premium_auto', 'months_since_last_claim',
       'months_since_policy_inception', 'number_of_open_complaints',
       'number_of_policies', 'policy_type', 'policy', 'renew_offer_type',
       'sales_channel', 'total_claim_amount', 'vehicle_class', 'vehicle_size',
       'vehicle_type'],
      dtype='object')

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [8]:
df1 = df[ (df["total_claim_amount"] > 1000) & (df["response"] == "Yes")]

In [9]:
df1["response"].unique()

array(['Yes'], dtype=object)

In [16]:
df1.sample(2)

unnamed:_0 customer       state  customer_lifetime_value response  \
8662        8662  YU86361  Washington             19160.989940      Yes   
4284        4284  GA72966      Oregon              8427.172486      Yes   

      coverage education effective_to_date employmentstatus gender  income  \
8662  Extended    Master           2/20/11         Employed      F   22705   
4284  Extended  Bachelor           2/10/11         Employed      F   46503   

     location_code marital_status  monthly_premium_auto  \
8662      Suburban       Divorced                   239   
4284      Suburban       Divorced                   215   

      months_since_last_claim  months_since_policy_inception  \
8662                     19.0                             59   
4284                      4.0                             21   

      number_of_open_complaints  number_of_policies    policy_type  \
8662                        0.0                   7  Personal Auto   
4284                        1.0                   1  Personal Auto   

           policy renew_offer_type sales_channel  total_claim_amount  \
8662  Personal L1           Offer2         Agent         1294.700423   
4284  Personal L3           Offer1         Agent         1032.000000   

     vehicle_class vehicle_size vehicle_type  
8662    Luxury SUV      Medsize            A  
4284    Luxury SUV      Medsize          NaN

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.


In [26]:
df_response_yes = df[df["response"] == "Yes"]

df2 = round(df_response_yes.groupby(["policy_type", "gender"])["total_claim_amount"].agg(["mean"]), 0)

df2

mean
policy_type    gender       
Corporate Auto F       434.0
               M       409.0
Personal Auto  F       453.0
               M       457.0
Special Auto   F       453.0
               M       430.0

Conclusions: On average females have a higher claim amount then males, apart from the policy type "Personal Auto" in which M and F have almost the same value. In between the types, the average amounts do not fluctuate very much, they are all between 409 and 457. The differences between F/M are greatest in Corporate, smaller in Special and only slightly different in Personal.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

Unfortunately, I dont understand this task, I have tried a couple of things (see below) but I dont think this was meant. 

In [94]:
df3 = df.groupby(["state", "policy"])["policy"].agg(["count"])
df3

count
state      policy             
Arizona    Corporate L1     62
           Corporate L2    119
           Corporate L3    200
           Personal L1     293
           Personal L2     434
           Personal L3     742
           Special L1       25
           Special L2       35
           Special L3       27
California Corporate L1    164
           Corporate L2    246
           Corporate L3    425
           Personal L1     439
           Personal L2     905
           Personal L3    1250
           Special L1       18
           Special L2       49
           Special L3       56
Nevada     Corporate L1     37
           Corporate L2     73
           Corporate L3    109
           Personal L1     136
           Personal L2     212
           Personal L3     391
           Special L1        5
           Special L2       18
           Special L3       12
Oregon     Corporate L1     94
           Corporate L2    174
           Corporate L3    324
           Personal L1     405
           Personal L2     618
           Personal L3    1157
           Special L1       21
           Special L2       63
           Special L3       53
Washington Corporate L1     40
           Corporate L2     46
           Corporate L3     89
           Personal L1     125
           Personal L2     230
           Personal L3     327
           Special L1        5
           Special L2       14
           Special L3       12

In [96]:
df.groupby("state").get_group("California")

unnamed:_0 customer       state  customer_lifetime_value response  \
1               1  KX64629  California              2228.525238       No   
6               6  IW72280  California              5035.035257       No   
7               7  IH64929  California              4956.247235       No   
8               8  FM55990  California              5989.773931      Yes   
13             13  KR82385  California              5454.587929       No   
...           ...      ...         ...                      ...      ...   
10895       10895  EG40670  California              6232.687923       No   
10898       10898  SO95845  California              5807.659307       No   
10899       10899  WU60905  California              8610.667489       No   
10902       10902  PP30874  California              3579.023825       No   
10908       10908  WA60547  California             11971.977650       No   

       coverage             education effective_to_date employmentstatus  \
1         Basic               College           1/18/11       Unemployed   
6         Basic                Doctor           2/14/11         Employed   
7         Basic  High School or Below           2/24/11         Employed   
8       Premium               College           1/19/11         Employed   
13        Basic                Master           1/26/11         Employed   
...         ...                   ...               ...              ...   
10895  Extended              Bachelor            1/8/11         Disabled   
10898     Basic               College            1/7/11         Employed   
10899     Basic  High School or Below            2/6/11       Unemployed   
10902  Extended  High School or Below           1/24/11         Employed   
10908   Premium               College           2/13/11         Employed   

      gender  income location_code marital_status  monthly_premium_auto  \
1          F       0      Suburban         Single                    64   
6          F   37405         Urban        Married                    63   
7          M   87197         Urban         Single                    63   
8          M   66839      Suburban         Single                   154   
13         M   66334      Suburban         Single                    69   
...      ...     ...           ...            ...                   ...   
10895      F   28334      Suburban         Single                    83   
10898      F   62447         Rural         Single                    74   
10899      F       0      Suburban        Married                   111   
10902      F   28304      Suburban        Married                    91   
10908      F   64195         Urban       Divorced                   158   

       months_since_last_claim  months_since_policy_inception  \
1                          3.0                             26   
6                          8.0                             99   
7                         35.0                             45   
8                         33.0                             24   
13                        25.0                             25   
...                        ...                            ...   
10895                      3.0                             44   
10898                      9.0                              2   
10899                     27.0                             36   
10902                     10.0                             30   
10908                      0.0                             27   

       number_of_open_complaints  number_of_policies     policy_type  \
1                            0.0                   1   Personal Auto   
6                            3.0                   4  Corporate Auto   
7                            0.0                   3   Personal Auto   
8                            0.0                   1   Personal Auto   
13                           0.0                   4   Personal Auto   
...                          ...                 ...             ..

In [78]:
df.state.value_counts()

state
California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: count, dtype: int64

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [102]:
round(df.groupby(["education", "gender"])["customer_lifetime_value"].agg(["max", "min", "median"]),2)

max      min   median
education            gender                            
Bachelor             F       73225.96  1904.00  5640.51
                     M       67907.27  1898.01  5548.03
College              F       61850.19  1898.68  5623.61
                     M       61134.68  1918.12  6005.85
Doctor               F       44856.11  2395.57  5332.46
                     M       32677.34  2267.60  5577.67
High School or Below F       55277.45  2144.92  6039.55
                     M       83325.38  1940.98  6286.73
Master               F       51016.07  2417.78  5729.86
                     M       50568.26  2272.31  5579.10

Conclusions:
General:
The customer_lifetime_value is highest regarding people with the lowest education level ("High School or Below") and lowest with the highest education level ("Doctor"). But in between, the data is not consistent (not in order: College, Bacherlor, Master), so the only clear correlation seems to be in the extremes.
Apart from the education level "Bachelor", where it is similar, males always generate more customer_lifetime_value in the median.
Some interesting findings:
The gender difference on education level "High School or Below" is significant, male make ~1/2 more than females. Not as big but also interesting is the gender difference in "Doctor" where females generates ~ 1/3 more than males.


## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here